File: backend/app/api/v1/schemas/verify.py
from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, field_validator


class CheckResult(BaseModel):
    status: str
    expected: Any = None
    actual: Any = None
    fix_suggestion: str | None = None
    dismissed_reason: str | None = None
    dismissed_by: str | None = None
    dismissed_at: str | None = None


class ApplyFixParameters(BaseModel):
    row_ids: list[int]


class ApplyFixRequest(BaseModel):
    report_id: UUID
    check_name: str
    fix_type: Literal["exclude_rows"]
    parameters: ApplyFixParameters


class DismissCheckRequest(BaseModel):
    report_id: UUID
    check_name: str
    reason: str

    @field_validator("reason")
    @classmethod
    def reason_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Reason must not be empty")
        return v.strip()


class AssumptionItem(BaseModel):
    text: str
    flag_type: str
    confidence: float
    source_reference: str


class AssumptionActionRequest(BaseModel):
    report_id: UUID
    assumption_index: int
    action: Literal["acknowledged", "signed_off", "rejected"]


class FigureTrace(BaseModel):
    figure_value: str
    source_rows: str
    formula: str
    match_status: str
    variance_pct: float


class VerifyResponse(BaseModel):
    report_id: str
    deck_id: str
    narrative_id: str
    passed: bool
    checks: dict[str, CheckResult]
    figure_traces: list[FigureTrace]
    assumptions: list[AssumptionItem] = []
    assumption_actions: list[dict[str, Any]] = []


class SourceRowsResponse(BaseModel):
    figure_value: str
    formula: str
    rows: list[dict[str, Any]]

File: frontend/src/types/verify.ts
export type FigureTrace = {
  figure_value: string;
  source_rows: string;
  formula: string;
  match_status: "exact" | "within_tolerance" | "mismatch";
  variance_pct: number;
};

export type CheckResult = {
  status: "pass" | "fail" | "dismissed";
  expected?: unknown;
  actual?: unknown;
  fix_suggestion?: string | null;
  dismissed_reason?: string;
  dismissed_by?: string;
  dismissed_at?: string;
};

export type ApplyFixRequest = {
  report_id: string;
  check_name: string;
  fix_type: "exclude_rows" | "recalculate";
  parameters: { row_ids: number[] };
};

export type DismissCheckRequest = {
  report_id: string;
  check_name: string;
  reason: string;
};

export type AssumptionItem = {
  text: string;
  flag_type: "EXPLICIT" | "PATTERN" | "INFERRED";
  confidence: number;
  source_reference: string;
};

export type AssumptionAction = {
  assumption_index: number;
  action: "acknowledged" | "signed_off" | "rejected";
  user_id: string;
  created_at: string;
};

export type AssumptionActionRequest = {
  report_id: string;
  assumption_index: number;
  action: "acknowledged" | "signed_off" | "rejected";
};

export type VerifyResponse = {
  report_id: string;
  deck_id: string;
  narrative_id: string;
  passed: boolean;
  checks: Record<string, CheckResult>;
  figure_traces: FigureTrace[];
  assumptions: AssumptionItem[];
  assumption_actions: AssumptionAction[];
  verified_at: string | null;
};

export type SourceRowsResponse = {
  figure_value: string;
  formula: string;
  rows: Record<string, unknown>[];
};
